pkg/storage/disk.go

```diff
@@ -19,44 +19,50 @@ const (
 )
 
 type vdiskModule struct {
-	v    pkg.VolumeAllocater
-	path string
+	module *Module
 }
 
 // NewVDiskModule creates a new disk allocator
-func NewVDiskModule(v pkg.VolumeAllocater) (pkg.VDiskModule, error) {
-	fs, err := v.Path(vdiskVolumeName)
-	if errors.Is(err, os.ErrNotExist) {
-		fs, err = v.CreateFilesystem(vdiskVolumeName, 0, pkg.SSDDevice)
-	}
+func NewVDiskModule(module *Module) (pkg.VDiskModule, error) {
+	return &vdiskModule{module: module}, nil
+}
 
+func (d *vdiskModule) findDisk(id string) (string, error) {
+	pools, err := d.module.VDiskPools()
 	if err != nil {
-		return nil, err
+		return "", errors.Wrapf(err, "failed to find disk with id '%s'", id)
+	}
+
+	for _, pool := range pools {
+		path, err := d.safePath(pool, id)
+		if err != nil {
+			return "", err
+		}
+
+		if _, err := os.Stat(path); err == nil {
+			// file exists
+			return path, nil
+		}
 	}
 
-	return &vdiskModule{v: v, path: filepath.Clean(fs.Path)}, nil
+	return "", os.ErrNotExist
 }
 
 // AllocateDisk with given size, return path to virtual disk (size in MB)
 func (d *vdiskModule) Allocate(id string, size int64) (string, error) {
-	path, err := d.safePath(id)
-	if err != nil {
-		return "", err
-	}
-
-	if _, err := os.Stat(path); err == nil {
-		// file exists
+	path, err := d.findDisk(id)
+	if err == nil {
 		return path, errors.Wrapf(os.ErrExist, "disk with id '%s' already exists", id)
 	}
 
-	supported, err := d.v.CanAllocate(vdiskVolumeName, uint64(size))
+	base, err := d.module.VDiskFindCandidate(uint64(size))
 	if err != nil {
-		return path, errors.Wrap(err, "failed to check capacity for this disk allocation")
+		return "", errors.Wrap(err, "failed to find a candidate to host vdisk")
```

Comment:
can we add the size to the error message

Reply:
Of course